scipy/signal/signaltools.py

```diff
@@ -892,41 +892,84 @@ def _prod(iterable):
     return product
 
 
-def _fftconv_faster(x, h, mode):
+def _conv_ops(x_shape, h_shape, mode):
+    x_size, h_size = _prod(x_shape), _prod(h_shape)
+    if mode == "full":
+        out_shape = [n + k - 1 for n, k in zip(x_shape, h_shape)]
+    elif mode == "valid":
+        out_shape = [max(n, k) - min(n, k) + 1 for n, k in zip(x_shape, h_shape)],
+    elif mode == "same":
+        out_shape = x_shape
+    else:
+        raise ValueError("Acceptable mode flags are 'valid',"
+                         " 'same', or 'full', not mode={}".format(mode))
+
+    s1, s2 = x_shape, h_shape
+    if len(x_shape) == 1:
+        s1, s2 = s1[0], s2[0]
+        if mode == "full":
+            direct_ops = s1 * s2
+        elif mode == "valid":
+            direct_ops = (s2 - s1 + 1) * s1 if s2 >= s1 else (s1 - s2 + 1) * s2
+        elif mode == "same":
+            direct_mul = s1 * s2 if s1 <= s2 else s1 * s2 - (s2 // 2) * ((s2 + 1) // 2)
+    else:
+        if mode == "full":
+            direct_ops = min(_prod(s1), _prod(s2)) * _prod(out_shape)
+        elif mode == "valid":
+            direct_ops = min(_prod(s1), _prod(s2)) * _prod(out_shape)
+        elif mode == "same":
+            direct_ops = _prod(s1) * _prod(s2)
+    fft_ops = sum(n * np.log(n) for n in (x_shape + h_shape + tuple(out_shape)))
```

Comment:
I think this will seriously under-estimate the cost of `fftconvolve` for higher dimensional input. With this formula, adding a new dimension adds a constant to the number of `fft_ops` but it should increase multiplicatively.

Reply:
I second this. The formula just happens to look ok for the region of parameters being tested - namely 1 dimensional, or 2 dimensional with small shapes.

For two identical shapes of `(n, n)`, this formula gives about `4nlog(n)`, while the correct number of multiplications is a constant away from `4n^2log(n)`. If the choices are a formula that we don't understand that predicts somewhat well, and a correct formula that doesn't predict well, then this project might not be fully baked yet.

Reply:
Whoops! Looks like this PR is lagging behind https://github.com/stsievert/scipy-convolve-timings/. In there, I'm using the code

``` python
    full_out_shape = [n + k - 1 for n, k in zip(S1, S2)]
    N = [_prod(shape) for shape in [S1, S2, full_out_shape]]
    fft_time = sum(n * np.log(n) for n in N)
```

(visible at https://github.com/stsievert/scipy-convolve-timings/blob/10ca85f255aa0032fb155c4df3d695393564a29a/common.py#L87-L88)

Reply:
Ah cools, sorry for the bother :)

I'm still getting something slightly different on paper. Can you explain the source of the terms corresponding to `S1` and `S2` in the line `N = [_prod(shape) for shape in [S1, S2, full_out_shape]]`?

Reply:
I've edited the comment above to make `S1 == x.shape` and `S2 == h.shape`.

Reply:
Yeah, that's cool. I'm trying to understand why the fft multiplications is a sum of these three terms. Say we have two identical shapes of `(n,)`. Then the sum here is `nlogn + nlogn + 2nlog(2n)`. Looking at the code, I see three ffts of size (approximately) `2n`, so I'm expecting only the last term in the sum.

The difference between the two formulas is going to show up as different O-constants when varying between equal sized shapes and very different size shapes. Specifically, when the shapes aren't identical, say `(n,)` and `(h,)`, with `h << n`, the current formula gives about `2nlogn`, while on paper I get `nlogn`.

So between these two edges - identical shapes, and very different shapes (one large, one small) - you won't be able to get a single correct O-constant.

Reply:
What code are you looking at?

Reply:
At `master` in `stsievert/scipy-convolve-timings`:
https://github.com/stsievert/scipy-convolve-timings/blob/master/common.py#L87

Reply:
Whoops, I meant for the statement "I see three ffts of size (approximately) 2n". What line are you seeing those FFTs on?

Reply:
Ahhhhhahaha ok
Here:
https://github.com/scipy/scipy/blob/2fea7fd1ab246be552ad2baf8fd4e60ee2929dbf/scipy/signal/signaltools.py#L382-L385

Reply:
(*) And I'm saying "approximately" because of the `next_fast_len` business

Reply:
https://github.com/scipy/scipy/blob/2fea7fd1ab246be552ad2baf8fd4e60ee2929dbf/scipy/signal/signaltools.py#L382-L385

where `fshape` is given by
https://github.com/scipy/scipy/blob/2fea7fd1ab246be552ad2baf8fd4e60ee2929dbf/scipy/signal/signaltools.py#L539-L540
(and a pass through `next_fast_len`)

Reply:
How the implemented expression incorrect? I'm seeing 3 FFT/IFFTs for shapes that are (approximately) `x_shape`, `h_shape` and `full_out_shape`, and performed in serial. The serial implementation indicates `sum` is the correct reduction.

> I see three ffts of size (approximately) 2n

One FFT is `sp1 = fft(in1, fshape, axes=axes)`. I think this does the recorded number of operations; [fftpack.fftn][1] does some padding/trimming to make the output shapes align after it does the FFT. The docstring says this about `fshape`:

> The shape of the result. ...  If `shape[i] > x.shape[i]`, the i-th dimension is padded with zeros. If `shape[i] < x.shape[i]`, the i-th dimension is truncated to length `shape[i]`.
> 
>—[fftpack.fftn docstring][1] 


[1]:https://docs.scipy.org/doc/scipy/reference/generated/scipy.fftpack.fftn.html#scipy.fftpack.fftn

Reply:
The `fshape` argument means that the ffts are of this shape, not the shape of the first argument. So all three ffts are on this same shape

Reply:
Ahhh I get the misunderstanding!

The padding/trimming happens **before** the FFT, not after :)

Reply:
You're incorrect, The padding is done _before_ taking the FFT. Otherwise the result would be incorrect.

Reply:
This is all just convention. In applications, it's very common to need the DFT at a sample rate of less or more than the shape of input, but handling the truncation/padding is tricky and the underlying implementation can also improve on memory allocation if allowed. So this option is baked into modern FFT apis - specifically, both in numpy and scipy.

But reading the doc string now, if I didn't know the convention beforehand, I'm not sure I would understand the api either. It really isn't clear when the truncation/padding happens.

Reply:
It's stated more explicitly in the [`fftpack.fft` docstring](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fftpack.fft.html#scipy.fftpack.fft) or in [`fft.fftn`'s docstring](https://scipy.github.io/devdocs/generated/scipy.fft.fftn.html#scipy.fft.fftn).